### DISCUSSION TITLE
Could libfive's core concepts be implemented by a non-expert?

### DISCUSSION THREAD

**Discussion by lf94** [libfive/libfive#520] [Q&A] [answered]
Hey Matt! Big love for libfive. I think it achieves some great things in the space.

My biggest question is do you think libfive could be re-implemented by a non-expert? A huge ideological thorn in my head is that if anyone is not able to re-implement a concept, no one should marry themselves to it. Self-sufficiency is big for me.

Could you please list the core concepts that someone would need to learn to implement a minimal version of libfive? Are all these concepts well explained?

Thank you very much for your great work!!

**Comment by doug-moen** [comment]
Are you looking for information on how to build a system that is like libfive Studio, that allows you to build and render 3D models described as signed distance functions? There are many projects like this, it's not that hard, and I've done it myself.

Or are you looking for information on how to recreate the specific and unique algorithm that Libfive uses to render a signed distance function as a triangle mesh? This AFAIK has only been implemented by Matt. Other projects use different algorithms that are well documented and have been successfully implemented by many people.

**Comment by lf94** [comment]
Hey Doug, hope all is well! I guess you've forgotten my username :sweat_smile:  Yeah yeah, building and rendering 3D models as SDFs we know is well within non-experts grasp. Yeah, the biggest is turning SDFs into triangle meshes. From what I understand libfive's algorithm is pretty much the best, but it seems difficult to understand. I was wondering if Matt could offer some advice here :slightly_smiling_face: 

(P.S. I'm hoping to see you pick up curv again!!)

**Comment by lf94** [comment]
@mkeeter Sorry, I should have started this as a discussion; bad habit.

**Comment by mkeeter** [comment, marked as answer]
This is a tough question to answer!  At one point, I was a non-expert, but I was able to implement `libfive`; it's therefore possible, but one of the steps along the way may be "become an expert".

The core algorithms are not _too tricky_: meshing is an implementation of [Manifold Dual Contouring](https://people.engr.tamu.edu/schaefer/research/dualsimp_tvcg.pdf), which requires some basic linear algebra but nothing exotic.  I expect that someone could implement the basics version of `libfive`'s meshing in 1-2K of Python (without tape simplification, i.e. given some `f(x, y, z) -> (v, dv/dx, dv/dy, dv/dz)`)

However, it would be much slower!

Most of the work has gone into making it **fast**, plus some amount of work into making it handle weird corner cases.  The biggest example of the latter is the "feature" handling, which solves [this problem](https://www.mattkeeter.com/blog/2017-05-13-implicit-zeros/).

(In retrospect, I'm not totally sure the latter is worth it, versus just declaring "zero-thickness boundaries may be meshed ambiguously")

**Reply by lf94** [reply]
Thanks for the reply!

Is there a definitive explanation of dual contouring?

I suspect making it fast is one of libfive's secret sauces. Really I see the tape simplification as the main innovation libfive brings along, so I think it'd be necessary to understand this.

> In retrospect, I'm not totally sure the latter is worth it, versus just declaring "zero-thickness boundaries may be meshed ambiguously"

Yeah, it'd probably be better to do usage recommendations than add complexity to the implementation (which could work against people in other unknown ways)...

**Reply by mkeeter** [reply]
> Is there a definitive explanation of dual contouring?

Dual Contouring is quite well documented, between the [original paper](https://www.cs.rice.edu/~jwarren/papers/dualcontour.pdf) and the ["Secret Sauce" write-up](https://people.eecs.berkeley.edu/~jrs/meshpapers/SchaeferWarren2.pdf).

> Really I see the tape simplification as the main innovation libfive brings along, so I think it'd be necessary to understand this.

Tape simplification is quite easy – it's [Algorithm 2 in this paper](https://www.mattkeeter.com/research/mpr/keeter_mpr20.pdf), which is only ~20 lines.

**Comment by ghost** [comment]


**Reply by lf94** [reply]
It's not about reinventing the wheel. It's about building the wheel at all.
